### DISCUSSION TITLE
GPOS handling in OpenJDK

### DISCUSSION THREAD

**Discussion by gredler** [harfbuzz/harfbuzz#3191] [Q&A] [answered]
Hi HarfBuzz team!

I'm working on diagnosing and fixing a Java (OpenJDK) bug where in some cases the vowels in Thai text are shifted left by one character, mangling the text [1]. The font that I'm using to replicate the issue, Google Noto Sans Thai Regular [2], uses the GPOS table internally. It looks like the GPOS adjustment in HB is a two-step process, where HB first sets the glyph position x_offset using anchor data from the GPOS table [3], and then adjusts that x_offset using existing x_advance data [4], all coordinated via the HB_BUFFER_SCRATCH_FLAG_HAS_GPOS_ATTACHMENT scratch flag.

Internally, OpenJDK is using the font_funcs virtual method functionality [5] to customize most of the font property callbacks [6]. Java provides users two ways to customize the size of a font: you can set the pt size, or you can set an affine transform (scale, rotate, shear, translate, etc). When you use the pt size approach, OpenJDK calls hb_font_set_scale with the appropriate pt size [7]. However, when a scaling affine transform is used, the pt size is technically unchanged, and a value of 1pt is used. Usually this works fine, because Java's font_funcs callbacks are providing scaled origin and advance numbers back to HB, but this breaks the two-step GPOS attachment logic: the first step is not aware of the scaling applied by the OpenJDK font_funcs callbacks (especially the hb_font_get_glyph_h_advance_func_t), but the second step uses a scaled advance value provided by an OpenJDK callback. In practice, at e.g. x50 scaling, this means that the first step sets the x_offset to a relatively small value (e.g. 32899), but the second step adjusts the x_offset by a relatively large value (e.g. 1612184).

Ideally the first GPOS step would be aware of the custom affine transform, so that both steps would be operating at the correct scale. One fix would be for OpenJDK to take the affine transform scale into account when setting the font size via hb_font_set_scale, but I'm not sure this would cover all scenarios, nor fit the OpenJDK design for affine transform handling.

I do wonder if there is a font_funcs design gap here, since there is no way to register a callback to customize the values of GPOS anchor / attachment points. If a HB user customizes all callbacks and applies custom logic to all position / advance values returned to HB, there is no way to ensure that similar logic is applied to GPOS values that are read directly by HB. Is there room for another callback which gives the user a hook to tweak GPOS values?

Looking forward to your feedback and ideas!

Daniel

[1] https://bugs.openjdk.java.net/browse/JDK-8269888
[2] https://www.google.com/get/noto/#sans-thai
[3] https://github.com/harfbuzz/harfbuzz/blob/bbeb3a62b0efbb598d8683f7c4b6cc7069a58aeb/src/hb-ot-layout-gpos-table.hh#L708
[4] https://github.com/harfbuzz/harfbuzz/blob/bbeb3a62b0efbb598d8683f7c4b6cc7069a58aeb/src/hb-ot-layout-gpos-table.hh#L2922
[5] https://harfbuzz.github.io/fonts-and-faces-custom-functions.html
[6] https://github.com/openjdk/jdk/blob/005d8a7fca8b4d9519d2bde0a7cdbbece1cd3981/src/java.desktop/share/native/libfontmanager/hb-jdk-font.cc#L271
[7] https://github.com/openjdk/jdk/blob/005d8a7fca8b4d9519d2bde0a7cdbbece1cd3981/src/java.desktop/share/native/libfontmanager/hb-jdk-font.cc#L422

**Comment by behdad** [comment, marked as answer]
Hi Daniel,

I believe this is a problem in how JDK implements HB font-funcs:

> Internally, OpenJDK is using the font_funcs virtual method functionality [5] to customize most of the font property callbacks [6]. Java provides users two ways to customize the size of a font: you can set the pt size, or you can set an affine transform (scale, rotate, shear, translate, etc). When you use the pt size approach, OpenJDK calls hb_font_set_scale with the appropriate pt size [7]. However, when a scaling affine transform is used, the pt size is technically unchanged, and a value of 1pt is used. Usually this works fine, because Java's font_funcs callbacks are providing scaled origin and advance numbers back to HB, but this breaks the two-step GPOS attachment logic: the first step is not aware of the scaling applied by the OpenJDK font_funcs callbacks (especially the hb_font_get_glyph_h_advance_func_t), but the second step uses a scaled advance value provided by an OpenJDK callback. In practice, at e.g. x50 scaling, this means that the first step sets the x_offset to a relatively small value (e.g. 32899), but the second step adjusts the x_offset by a relatively large value (e.g. 1612184).

That's exactly the problem.

> Ideally the first GPOS step would be aware of the custom affine transform, so that both steps would be operating at the correct scale. One fix would be for OpenJDK to take the affine transform scale into account when setting the font size via hb_font_set_scale, but I'm not sure this would cover all scenarios, nor fit the OpenJDK design for affine transform handling.

That is indeed the correct fix.

> I do wonder if there is a font_funcs design gap here, since there is no way to register a callback to customize the values of GPOS anchor / attachment points. If a HB user customizes all callbacks and applies custom logic to all position / advance values returned to HB, there is no way to ensure that similar logic is applied to GPOS values that are read directly by HB. Is there room for another callback which gives the user a hook to tweak GPOS values?

No. `set_scale` is indeed what you need to make sure you call correctly. As you noticed, it doesn't take a full affine. There's an assumption that horizontal typesetting happens in horizontal direction, and vertical typesetting happens in vertical direction. So we don't support full affine. We can possible add `set_skew` to support skews better, but that's separate.

Another fix is to always apply the affine yourself. That's how `hb-ft` works I think.

Back in 2013, I noticed a similar problem with how the JDK fonts work, when I was writing `icu-le-hb`:
https://github.com/harfbuzz/icu-le-hb/blob/04505b2b7280197150c27332a71d8b232272493a/src/LayoutEngine.cpp#L182-L197

Hope that helps.

**Reply by gredler** [reply]
Thanks for the pointers, very helpful!
